theano/compile/function.py

```diff
@@ -185,6 +185,22 @@ def opt_log1p(node):
 
 
     """
+    if isinstance(outputs, dict):
+        output_items = outputs.items()
+
+        output_items_sorted = sorted(output_items)
```

Comment:
My worry is that the behaviour of `sorted` is not deterministic between runs.
I think we should restrict it to strings, at least for python 2 (maybe python 3 will raise an exception if the objects are not directly comparable).